### PR TITLE
Prioritize high confidence stats during broadcast joins

### PR DESCRIPTION
**Pull request by abhinavmuk04**
## Description
Prioritize high confidence stats during broadcast joins if enabled

## Motivation and Context
When there are two PlanNodes in which they are both small enough for broadcast join we will prioritize the side which has higher confidence stats. If they both have high confidence stats then we keep the original behavior. The user has the ability to turn this on and off.

## Impact
This change will create a feature which the user can utilize to improve optimization and help improve the execution time of broadcast join queries

## Test Plan
Implemented various tests in both DetermineJoinDistributionType and ReorderJoinsType, which will check if, with the session property enable, nodes with the higher confidence stats will be broadcasted

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes
```
General Changes
* Add confidence based broadcasting, side of join with highest confidence will be on build side. 
  This can be enabled with the ``confidence_based_broadcast`` session property :pr:`23016`
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:255]: 1. confidenceLevel high does not necessarily mean it's hbo (e.g. could be table scan stats or some other thing with high confidence that gets added in the future), so the method name doesn't accurately reflect what this method is checking
2. That being said, we don't care whether the stats are from hbo. We just want to choose the stats with the highest confidence.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:545]: It would be better to call into DetermineJoinDistributionType here so as not to repeat code.  Also, that would enable us to incorporate other features of DetermineJoinDistributionType (e.g. size based join).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:128]: consider how to incorporate this with the size based join paths.  I also feel that this rule now has too many branches for how we choose which table to broadcast (not just from your change, there just keep being more heurisitics added) and you end up in these mutually exclusive branches where either you need to repeat code to e.g. choose the highest confidence stats from multiple branches of the code, or you only get that heuristic if you go down one branch but not the other. It would be nice to consolidate them.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:149]: CONFIDENCE_BASED_BROADCASTING_ENABLED -> CONFIDENCE_BASED_BROADCAST_ENABLED

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:154]: Return an optional field to avoid returning null

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:158]: I was thinking that, FACT > LOW/HIGH and HIGH > LOW, not just at least HIGH

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:182]: Unexpected change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:543]: If this function is referenced in multiple files, consider find a util file for this function.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:157]: I personally feel that this confidenceBasedBroadcastingEnabled change should not belong to this function, but the caller of this function.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java
Comment (feilong-liu) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java:265]: I didn't see the setConfidence() here, I guess the default is LOW?

File: presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java
Comment (feilong-liu) [inline presto-tests/src/test/java/com/facebook/presto/execution/TestHistoryBasedStatsTracking.java:500]: You didn't enable CONFIDENCE_BASED_BROADCASTING_ENABLED in this test?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java
Comment (abhinavmuk04) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java:265]: @feilong-liu Yes, default confidence is LOW, but to make it more clear I can set the confidence as LOW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (abhinavmuk04) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:158]: @feilong-liu I thought about this as well. But I was thinking if there was something with High Confidence with a lower estimate than something with Fact Confidence, wouldn't it make sense to pick the smaller one?

Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: High confidence can still be inaccurate, fact is always true

Comment (abhinavmuk04) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: Okay that makes sense, I'll go ahead and make that change

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/ConfidenceBasedBroadcast.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/ConfidenceBasedBroadcast.java:25]: Rename the class with Util in name to make it different from regular classes

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:131]: combine the if with the above if?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:167]: change is still here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:542]: You need to check whether the feature is enabled or not

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/ConfidenceBasedBroadcastUtil.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/ConfidenceBasedBroadcastUtil.java:35]: Instead of doing this multiple if check, can you just do comparison on the confidence level and pick the side with higher confidence to be build side, which should be functionally the same but more concise.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java:543]: Did you forget the must partition check?

**Review by feilong-liu — APPROVED** [review]

**Comment by feilong-liu** [comment]
Code lgtm. However, as a code owner, I do not have ownership for the SystemSessionProperty file, will need an committer approval for help to merge the change here.

**Review by NikhilCollooru — APPROVED** [review]

**Pull request merged**
